cmake/ranges_flags.cmake

```diff
@@ -0,0 +1,173 @@
+  endif()
+  if (RANGES_ENV_MACOSX)
+    ranges_append_flag(RANGES_HAS_NO_IMPLICIT_MODULE_MAPS -fno-implicit-module-maps)
+  endif()
+  if (RANGES_DEBUG_BUILD)
+    ranges_append_flag(RANGES_HAS_GMODULES -gmodules)
+  endif()
+endif()
+
+# Sanitizer support: detect incompatible sanitizer combinations
+if (RANGES_ASAN AND RANGES_MSAN)
+  message(FATAL_ERROR "[range-v3 error]: AddressSanitizer and MemorySanitizer are both enabled at the same time!")
+endif()
+
+if (RANGES_MSAN AND RANGES_ENV_MACOSX)
+  message(FATAL_ERROR "[range-v3 error]: MemorySanitizer is not supported on MacOSX!")
+endif()
+
+# AddressSanitizer support
+if (RANGES_ASAN)
+  if (RANGES_ENV_MACOSX) # LeakSanitizer not supported on MacOSX
+    ranges_append_flag(RANGES_HAS_SANITIZE_FLAGS "-fsanitize=address,integer,undefined")
+  else()
+    ranges_append_flag(RANGES_HAS_SANITIZE_FLAGS "-fsanitize=address,integer,undefined, leak")
+  endif()
+  ranges_append_flag(RANGES_HAS_SANITIZE_NO_RECOVER "-fno-sanitize-recover=address,integer,undefined,leak")
+  ranges_append_flag(RANGES_HAS_NO_OMIT_FRAME_POINTER -fno-omit-frame-pointer)
+endif()
+
+# MemorySanitizer support
+if (RANGES_MSAN)
+  ranges_append_flag(RANGES_HAS_MSAN "-fsanitize=memory")
+  ranges_append_flag(RANGES_HAS_SANITIZE_RECOVER_ALL "-fsanitize-recover=all")
+  ranges_append_flag(RANGES_HAS_NO_OMIT_FRAME_POINTER -fno-omit-frame-pointer)
+endif()
+
+# Build types:
+if (RANGES_DEBUG_BUILD AND RANGES_RELEASE_BUILD)
+  message(FATAL_ERROR "[range-v3 error] Cannot simultaneously generate debug and release builds!")
+endif()
+
+if (RANGES_DEBUG_BUILD)
+  ranges_append_flag(RANGES_HAS_O0 -O0)
+  ranges_append_flag(RANGES_HAS_NO_INLINE -fno-inline)
+  ranges_append_flag(RANGES_HAS_STACK_PROTECTOR_ALL -fstack-protector-all)
+  ranges_append_flag(RANGES_HAS_NO_STRICT_ALIASING -fno-strict-aliasing)
+  ranges_append_flag(RANGES_HAS_G3 -g3)
+  # Clang can generate debug info tuned for LLDB or GDB
+  if (RANGES_CXX_COMPILER_CLANG) 
+    if (RANGES_ENV_MACOSX)
+      ranges_append_flag(RANGES_HAS_GLLDB -glldb)
+    elseif(RANGES_ENV_LINUX)
+      ranges_append_flag(RANGES_HAS_GLLDB -gddb)
+    endif()
+  endif()
+endif()
+
+if (RANGES_RELEASE_BUILD)
+  if (NOT RANGES_ASSERTIONS)
+    ranges_append_flag(RANGES_HAS_DNDEBUG -DNDEBUG)
+  endif()
+  if (NOT RANGES_ASAN AND NOT RANGES_MSAN)
+    # The quality of ASan and MSan error messages suffers if we disable the
+    # frame pointer, so leave it enabled when compiling with either of them:
+    ranges_append_flag(RANGES_HAS_OMIT_FRAME_POINTER -fomit-frame-pointer)
+  endif()
+  
+  ranges_append_flag(RANGES_HAS_MARCH_NATIVE "-march=native")
+  ranges_append_flag(RANGES_HAS_MTUNE_NATIVE "-mtune=native")
+  ranges_append_flag(RANGES_HAS_OFAST -Ofast)
+
+  ranges_append_flag(RANGES_HAS_STRICT_ALIASING -fstrict-aliasing)
+  ranges_append_flag(RANGES_HAS_STRICT_VTABKE_POINTERS -fstrict-vtable-pointers)
```

Comment:
fixed, I've also enabled the nullability checks, and make all checks terminate the program by default so that we see the builds failing on travis.